lib/puppet/pops/types/string_converter.rb

```diff
@@ -0,0 +1,923 @@
+      result
+    end
+
+    # Merges two formats where the `higher` format overrides the `lower`. Produces a new `Format`
+    # @param [Format] lower
+    # @param [Format] higher
+    # @returns [Format] the merged result
+    #
+    def self.merge(lower, higher)
+      unless lower && higher
+        return lower || higher
+      end
+      lower.merge(higher)
+    end
+
+    # Merges a type => format association and returns a new merged and sorted association.
+    # @param [Format] lower
+    # @param [Format] higher
+    # @returns [Hash] the merged type => format result
+    #
+    def self.merge_string_formats(lower, higher)
+      unless lower && higher
+        return lower || higher
+      end
+      merged = (lower.keys + higher.keys).uniq.map do |k|
+        [k, merge(lower[k], higher[k])]
+      end
+      sort_formats(merged)
+    end
+
+    # Sorts format based on generality of types - most specific types before general
+    #
+    def self.sort_formats(format_map)
+      format_map = format_map.sort do |(a,_),(b,_)| 
+        ab = b.assignable?(a)
+        ba = a.assignable?(b)
+        if a == b
+          0
+        elsif ab && !ba
+          -1
+        elsif !ab && ba
+          1
+        else
+          # arbitrary order if disjunct (based on name of type)
+          rank_a = type_rank(a)
+          rank_b = type_rank(b)
+          if rank_a == 0 || rank_b == 0
+            a.to_s <=> b.to_s
+          else
+            rank_a <=> rank_b
+          end
+        end
+      end
+      format_map.to_h
+    end
+
+    # Ranks type on specificity where it matters
+    # lower number means more specific
+    def self.type_rank(t)
+      case t
+      when PStructType
+        1
+      when PHashType
+        2
+      when PTupleType
+        3
+      when PArrayType
+        4
+      when PPatternType
+        10
+      when PEnumType
+        11
+      when PStringType
+        12
+      else
+        0
+      end
+    end
+    # Returns an array with a delimiter pair derived from the format.
+    # If format does not contain a delimiter specification the given default is returned
+    # 
+    # @param [Array<String>] the default delimiters
+    # @returns [Array<String>] a tuple with left, right delimiters
+    #
+    def delimiter_pair(default = StringConverter::DEFAULT_ARRAY_DELIMITERS)
+      DELIMITER_MAP[ @delimiters || @plus ] || default
+    end
+
+    def to_s
+      "%#{@flags}#{@width}.#{@prec}#{@format}"
+    end
+  end
+
+  # @api public
+  def self.convert(value, string_formats = :default)
+    singleton.convert(value, string_formats)
+  end
+
+  # @return [TypeConverter] the singleton instance
+  #
+  # @api public
+  def self.singleton
+    @tconv_instance ||= new
+  end
+
+  # @api private
+  #
+  def initialize
+    @@string_visitor   ||= Visitor.new(self, "string", 3, 3)
+    @@type_calculator = TypeCalculator.singleton()
```

Comment:
This is not needed. All occurrences of {{@@type_calculator.infer_set}} can be replaced with {{TypeCalculator.infer_set}} (it's implemented on the class).